### PR TITLE
[onert] Remove duplicated codes in base_loader

### PR DESCRIPTION
**Pull request by ragmani**
For issue #3625

This commit removes duplicated codes in base_loader.
  - Remove a template parameter from class BaseLoader
  - Unify loading operation to using loadOperationTo()

Signed-off-by: ragmani <ragmani0216@gmail.com>

### CODE REVIEW

**Review by wateret — DISMISSED** [review]
LGTM

**Comment by wateret** [comment]
It may be nice to request review to the original author of base_loader(I don't know who 😅 ) too.

**Comment by ragmani** [comment]
> It may be nice to request review to the original author of base_loader(I don't know who sweat_smile ) too.

@s-barannikov Could you review this PR?

**Review by wateret — APPROVED** [review]
LGTM

**Review by s-barannikov — COMMENTED** [review]
It seems that the PR mixes refactoring (e.g. inlining load* methods and introducing `loadOprationTo`) and changing logic (e.g. removing template parameter and introducing virtual method).
It would be much easier to review if the refactoring part is merged separately, or at least in a separate commit(s). Is it possible to split the changes? It would be really helpful.

File: runtime/onert/frontend/base_loader/include/base_loader.h
Comment (s-barannikov) [inline runtime/onert/frontend/base_loader/include/base_loader.h:74]: (historical note)
It was intended to use static polymorphism (via SpecificLoader) instead of the dynamic one (via virtual methods).
The workflow suggested was as follows:
- Initially, all the code is contained in BaseLoader.
- When Circle implementation of a specific concept begins to differ from the TFLite counterpart (as happened with SubGraphs and some Circle-specific operators), the related code is completely removed from the base loader and implemented in both derived loaders.
- When there become too many differences between the two loaders, the base loader can be removed, leaving two completely independent loaders (Circle and TFLite).
---
I must admit that the intention is not clear from the code, so the implementation of the idea might not be the best.

Virtual methods can also help to achieve the same goal. Applying virtual methods for such purposes is known as "template method pattern" (although it has nothing to do with C++ templates, it's just a name clash). It is well known pattern, so the code using it should be more clear to developers than the former approach using static polymorphism.

Please note that the virtual function should be `private` in this case (in both base and derived classes). More on this [here](http://www.gotw.ca/publications/mill18.htm).

PS
It is worth noting that each virtual method added to BaseLoader is actually two methods (one for each loader). This is because TFLiteLoader and CircleLoader derive from different base classes, thanks to template parameter of BaseLoader.

File: runtime/onert/frontend/base_loader/include/base_loader.h
Comment (s-barannikov) [inline runtime/onert/frontend/base_loader/include/base_loader.h:131]: I think it is ok to have some methods to keep the `switch` smaller.

File: runtime/onert/frontend/base_loader/include/base_loader.h
Comment (s-barannikov) [inline runtime/onert/frontend/base_loader/include/base_loader.h:1264]: Is this commented out intentionally?

File: runtime/onert/frontend/base_loader/include/base_loader.h
Comment (s-barannikov) [inline runtime/onert/frontend/base_loader/include/base_loader.h:1001]: I understand that this is cut & paste, but as it is a refactoring, I would suggest to remove such obvious comments.

File: runtime/onert/frontend/base_loader/include/base_loader.h
Comment (s-barannikov) [inline runtime/onert/frontend/base_loader/include/base_loader.h:907]: This is a great helper that reduces amount of boilerplate significantly. Why don't submit it (and its uses) first? This would make the reviewing much easier...

**Comment by s-barannikov** [comment]
Looks good (aside from minor comments), but please try to split this into smaller chunks.

**Comment by glistening** [comment]
+1 for https://github.com/Samsung/ONE/pull/4016#pullrequestreview-476977366. I've experienced a regression from a big change like this PR, which prevent careful review.

File: runtime/onert/frontend/base_loader/include/base_loader.h
Comment (hseok-oh) [reply runtime/onert/frontend/base_loader/include/base_loader.h:131]: I agree with @s-barannikov comment. I think `switch` must smaller and simple pattern as possible.

**Comment by hseok-oh** [comment]
Some part agree, and some part disagree.

Agree
- Introduce `loadOperationTo()`
- Introduce virtual method and remove typename of template class `BaseLoader`

Disagree
- Inlining parameter parsing in switch-case. I prefer to use simple switch-case just calling one method, then call common method in operator specific method.

File: runtime/onert/frontend/base_loader/include/base_loader.h
Comment (ragmani) [reply runtime/onert/frontend/base_loader/include/base_loader.h:1001]: Okay I will remove it.

File: runtime/onert/frontend/base_loader/include/base_loader.h
Comment (ragmani) [reply runtime/onert/frontend/base_loader/include/base_loader.h:1264]: This comment already existed in the original code. I don't know why this didn't commented out in detail. But It seems unnecessary. I will remove it.

File: runtime/onert/frontend/base_loader/include/base_loader.h
Comment (ragmani) [reply runtime/onert/frontend/base_loader/include/base_loader.h:131]: I think these operation's IR should be changed some params to input that can be tensor of backend. After it, the duplicated codes can be removed in this method. I will change this commend from removing these methods to remove duplicated codes in these methods.

**Comment by ragmani** [comment]
> Inlining parameter parsing in switch-case. I prefer to use simple switch-case just calling one method, then call common method in operator specific method.

I see. I will keep some method for simple switch-case.

File: runtime/onert/frontend/base_loader/include/base_loader.h
Comment (ragmani) [reply runtime/onert/frontend/base_loader/include/base_loader.h:74]: Thanks for your kindly explanation. I will move the virtual function to be `private`.

**Comment by ragmani** [comment]
> It seems that the PR mixes refactoring (e.g. inlining load* methods and introducing loadOprationTo) and changing logic (e.g. removing template parameter and introducing virtual method).

Right. I will change this PR to the two separated commit within this PR.

File: runtime/onert/frontend/base_loader/include/base_loader.h
Comment (ragmani) [reply runtime/onert/frontend/base_loader/include/base_loader.h]: I divided this PR to two separated commits.

File: runtime/onert/frontend/circle/src/circle_loader.cc
Comment (s-barannikov) [inline runtime/onert/frontend/circle/src/circle_loader.cc:94]: Should be private

File: runtime/onert/frontend/tflite/src/tflite_loader.cc
Comment (s-barannikov) [inline runtime/onert/frontend/tflite/src/tflite_loader.cc:81]: Same

**Review by s-barannikov — APPROVED** [review]
LGTM

File: runtime/onert/frontend/circle/src/circle_loader.cc
Comment (ragmani) [reply runtime/onert/frontend/circle/src/circle_loader.cc:94]: Thanks for your pointing out. It's my mistake.

File: runtime/onert/frontend/tflite/src/tflite_loader.cc
Comment (ragmani) [reply runtime/onert/frontend/tflite/src/tflite_loader.cc:81]: Thanks.

**Review by s-barannikov — APPROVED** [review]

**Review by hseok-oh — APPROVED** [review]

**Review by glistening — APPROVED** [review]
LGTM

**Comment by ragmani** [comment]
PTAL @wateret

**Comment by hseok-oh** [comment]
It's better to merge after branch out release 1.9.0, IMO.

**Comment by ragmani** [comment]
> It's better to merge after branch out release 1.9.0, IMO.

@hseok-oh Okay

**Comment by wateret** [comment]
Will merge once branch 1.9.0 is created.

**Comment by hseok-oh** [comment]
Release 1.9.0 branch is created.
@wateret PTAL

**Pull request merged**
